### PR TITLE
using gpg private key to sign the db archive

### PR DESCRIPTION
**Pull request by odra**
Uses gpg private key and password to encrypt and sign the database archive.

### CODE REVIEW

File: mongodb/README.md [hunk 1]
```diff
@@ -14,25 +14,27 @@ The job template requires a number of parameters to be specified. These can be v
...
+MONGODB_PASSWORD                  MongoDB user password                                                                     
 MONGODB_AUTHENTICATION_DATABASE   MongoDB database to authenticate against                                                  admin
-GPG_RECIPIENT                     GPG recpient name to be used to encrypt the database archive
-GPG_PUBLIC_KEY                    GPG public key content (base64 encoded)
-GPG_TRUST_MODEL                   GPG encryption trust model, defaults to "always"
+GPG_RECIPIENT                     GPG recpient name to be used to encrypt the database archive                              
+GPG_PUBLIC_KEY                    GPG public key content (base64 encoded)                                                   
+GPG_PRIVATE_KEY                   GPG private key content (base64 encoded)                                                  
+GPG_TRUST_MODEL                   GPG encryption trust model, defaults to "always"                                          always
+GPG_PASSWORD                      GPG password to be used when signing the backup archive
```
Comment (pmccarthy) [inline mongodb/README.md]: @odra AFAIK to encrypt a file using GPG we shouldn't need the password or private key. These should only be required when decrypting the file? The public key is all that's required when encrypting.

Example:
```
gpg -r <recipient> --yes --trust-model always --encrypt <filename>
```

Comment (odra) [reply mongodb/README.md]: I got an "user not found" issue when doing it without the private key but I believe that is due to the "---sign" param, let me remove the sign stuff and try it again.

**Comment by odra** [comment]
@pmccarthy I've updated the pr

File: mongodb/mongodb-backup-s3-job-template.yaml [hunk 2]
```diff
@@ -42,6 +42,8 @@ objects:
                   value: ${AWS_SECRET_ACCESS_KEY}
                 - name: GPG_RECIPIENT
                   value: ${GPG_RECIPIENT}
+                - name: GPG_PASSWORD
```
Comment (pmccarthy) [inline mongodb/mongodb-backup-s3-job-template.yaml]: @odra Is this required now with updated changes?

**Comment by pmccarthy** [comment]
@odra Tested Mongo job but backup files are still not being encrypted. Standard gz files in the S3 bucket.

```
oc new-app mongodb-backup-s3-job-template.yaml -p AWS_ACCESS_KEY_ID=XXXX -p AWS_SECRET_ACCESS_KEY=XXXX-p AWS_S3_BUCKET_NAME=rhm-eng-a-system-eu-west-1 -p MONGODB_USER=admin -p MONGODB_PASSWORD=XXXX -p MONGODB_HOST=mongodb-1 -p MONGODB_AUTHENTICATION_DATABASE=admin -p GPG_RECIPIENT=pamccart@redhat.com -p "GPG_PUBLIC_KEY=$(cat keys_public.gpg | base64)"
```

**Comment by pmccarthy** [comment]
@odra It looks like the files have been encrypted correctly as I can see them in the S3 bucket, have downloaded them and decrypted them using the set passphrase. I did notice, however, the following warning message for each backup?

```
WARNING: fh-reporting-13:51:38.dump.gz.gpg: Owner username not known. Storing UID=1000590000 instead.
```

File: scripts/mongodb/backup-s3.sh [hunk 3]
```diff
@@ -15,21 +15,29 @@ DATESTAMP=$(date '+%Y-%m-%d')
...
+# Imports gpg keys
 gpg --import /opt/rh/secrets/gpg_public_key
+gpg --list-keys
 
 # For each database archive data and push directly to S3
 for DATABASE in $DATABASES; do
   TIMESTAMP=$(date '+%H:%M:%S')
+  echo "==> Archiving database \"$DATABASE\""
+  mongodump -h $HOST:$PORT -u $USER -p $PASSWORD -d $DATABASE --archive --gzip --authenticationDatabase $AUTH_DB > /tmp/$DATABASE-$TIMESTAMP.dump.gz
+  echo "==> Enncrypting database archive \"$DATABASE\""
```
Comment (pmccarthy) [inline scripts/mongodb/backup-s3.sh]: Slight typo here with two n's in the word Encrypting. Normally wouldn't be concerned with this but as the backup logs may be customer facing it should be fixed.

**Comment by odra** [comment]
I assumed this message (`Owner username not known. Storing UID=1000590000 instead.`) was due to the docker container user being used by openshift

**Comment by jessesarn** [comment]
Its safe to ignore.  Had the tar contents been actually extracted to a filesystem, the UID/GID would be retained that existed originally.  Since we only deal with streams in and out, the owner of the file never really matters.

**Comment by odra** [comment]
After some debugging this is the file info inside the container:

```
sh-4.2$ stat /tmp/admin-15\:00\:51.dump.gz
  File: '/tmp/admin-15:00:51.dump.gz'
  Size: 846       	Blocks: 8          IO Block: 4096   regular file
Device: 90h/144d	Inode: 176         Links: 1
Access: (0644/-rw-r--r--)  Uid: (1000150000/ UNKNOWN)   Gid: (    0/    root)
Access: 2018-04-30 15:00:52.305958379 +0000
Modify: 2018-04-30 15:00:52.305958379 +0000
Change: 2018-04-30 15:00:52.305958379 +0000
 Birth: -
sh-4.2$ stat /tmp/admin-15\:00\:51.dump.gz.gpg 
  File: '/tmp/admin-15:00:51.dump.gz.gpg'
  Size: 1192      	Blocks: 8          IO Block: 4096   regular file
Device: 90h/144d	Inode: 179         Links: 1
Access: (0644/-rw-r--r--)  Uid: (1000150000/ UNKNOWN)   Gid: (    0/    root)
Access: 2018-04-30 15:00:52.555957517 +0000
Modify: 2018-04-30 15:00:52.305958379 +0000
Change: 2018-04-30 15:00:52.305958379 +0000
 Birth: -
```

The container itself is not using a specific username, just the UID.

**Review by pmccarthy — APPROVED** [review]
Re-ran verification steps and all backups were encrypted in the S3 bucket as expected. I also downloaded and decrypted a backup file successfully.

Looks good @odra  👍

**Pull request merged**